vertx-service-discovery/src/main/java/io/vertx/servicediscovery/impl/DefaultServiceDiscoveryBackend.java

```diff
@@ -46,15 +54,33 @@ public void store(Record record, Handler<AsyncResult<Record>> resultHandler) {
     if (record.getRegistration() != null) {
       throw new IllegalArgumentException("The record has already been registered");
     }
-
     record.setRegistration(uuid);
-    registry.put(uuid, record.toJson().encode(), ar -> {
-      if (ar.succeeded()) {
-        resultHandler.handle(Future.succeededFuture(record));
-      } else {
-        resultHandler.handle(Future.failedFuture(ar.cause()));
-      }
-    });
+    retrieveRegistry(registry ->
+      registry.put(uuid, record.toJson().encode(), ar -> {
+        if (ar.succeeded()) {
+          resultHandler.handle(Future.succeededFuture(record));
+        } else {
+          resultHandler.handle(Future.failedFuture(ar.cause()));
+        }
+      })
+    );
+  }
+
+  private synchronized void retrieveRegistry(Handler<AsyncMap<String, String>> handler) {
```

Comment:
You should use a handler of `AsyncResult` to make it possible to propagate any failure.